tests/test_new_return_types_qnode.py

```diff
@@ -0,0 +1,1665 @@
+        assert all(isinstance(r, tuple) for r in res)
+        assert all(isinstance(measurement_res[0], np.ndarray) for measurement_res in res)
+        assert all(isinstance(measurement_res[1], dict) for measurement_res in res)
+
+        for idx, shots in enumerate(raw_shot_vector):
+            for i, r in enumerate(res[idx]):
+                num_wires = len(sample_wires)
+                if shots == 1 and i % 2 == 0:
+                    expected_shape = () if num_wires == 1 else (num_wires,)
+                    assert r.shape == expected_shape
+                elif i % 2 == 0:
+                    expected_shape = (shots,) if num_wires == 1 else (shots, num_wires)
+                    assert r.shape == expected_shape
+                else:
+                    assert isinstance(r, dict)
+
+    @pytest.mark.parametrize("meas1,meas2", scalar_probs_multi)
+    def test_scalar_probs_sample_counts(self, shot_vector, meas1, meas2, device):
+        """Test scalar-valued, probability, sample and counts measurements all
+        in a single qfunc."""
+        qml.enable_return()
+
+        dev = qml.device(device, wires=5, shots=shot_vector)
+        raw_shot_vector = [
+            shot_tuple.shots for shot_tuple in dev.shot_vector for _ in range(shot_tuple.copies)
+        ]
+
+        def circuit(x):
+            qml.Hadamard(wires=[0])
+            qml.CRX(x, wires=[0, 1])
+            return (
+                qml.apply(meas1),
+                qml.apply(meas2),
+                qml.sample(qml.PauliX(4)),
+                qml.counts(qml.PauliX(3)),
+            )
+
+        qnode = qml.QNode(circuit, dev, diff_method=None)
+
+        res = qnode(0.5)
+
+        qml.disable_return()
+
+        all_shots = sum([shot_tuple.copies for shot_tuple in dev.shot_vector])
+
+        assert isinstance(res, tuple)
+        assert len(res) == all_shots
+        assert all(isinstance(r, tuple) for r in res)
+
+        for res_idx, meas_res in enumerate(res):
+            for i, r in enumerate(meas_res):
+                num_meas = i % 4
+                expval_or_var = num_meas == 0
+                probs = num_meas == 1
+                sample = num_meas == 2
+
+                if expval_or_var:
+                    assert r.shape == ()
+                elif probs:
+                    assert r.shape == (2**2,)
+
+                    # Probs add up to 1
+                    assert np.allclose(sum(r), 1)
+                elif sample:
+                    shots = raw_shot_vector[res_idx]
+                    if shots == 1:
+                        assert r.shape == ()
+                    else:
+                        expected = (shots,)
+                        assert r.shape == expected
+                else:
+                    # Return is Counts
+                    assert isinstance(r, dict)
+
+
+class TestIntegrationJacobianBackpropMultipleReturns:
+    """Test the new return types for the Jacobian of multiple measurements, with backprop."""
+
```

Comment:
It's great that backprop seems to be working for all interfaces 😄. Would we also want to test the forward pass for each interface then?

Reply:
I added them we have +1500 lines of test :')